contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,657 @@
+        if (oracle.getConsensusContract() != address(hashConsensus)) {
+            revert IncorrectOracleAndHashConsensusBinding(address(oracle), address(hashConsensus));
+        }
+        // TODO: check the binding in opposite direction when the view is added to HashConsensus
+    }
+
+    function _checkContractVersions() internal view {
+        _assertContractVersion(_lido(), EXPECTED_FINAL_LIDO_VERSION);
+        _assertContractVersion(_nodeOperatorsRegistry, EXPECTED_FINAL_NODE_OPERATORS_REGISTRY_VERSION);
+        _assertContractVersion(_legacyOracle(), EXPECTED_FINAL_LEGACY_ORACLE_VERSION);
+        _assertContractVersion(_accountingOracle(), EXPECTED_FINAL_ACCOUNTING_ORACLE_VERSION);
+        _assertContractVersion(_stakingRouter(), EXPECTED_FINAL_STAKING_ROUTER_VERSION);
+        _assertContractVersion(_validatorsExitBusOracle(), EXPECTED_FINAL_VALIDATORS_EXIT_BUS_ORACLE_VERSION);
+        _assertContractVersion(_withdrawalQueue(), EXPECTED_FINAL_WITHDRAWAL_QUEUE_VERSION);
+        _assertContractVersion(_withdrawalVault(), EXPECTED_FINAL_WITHDRAWAL_VAULT_VERSION);
+    }
+
+    function _assertContractVersion(IVersioned versioned, uint256 expectedVersion) internal view {
+        if (versioned.getContractVersion() != expectedVersion) {
+            revert InvalidContractVersion(address(versioned), expectedVersion);
+        }
+    }
+
+    function _transferOZAdminFromThisToAgent(IAccessControlEnumerable accessControlled) internal {
+        accessControlled.grantRole(DEFAULT_ADMIN_ROLE, _agent);
+        accessControlled.renounceRole(DEFAULT_ADMIN_ROLE, address(this));
+    }
+
+    function _resumePausableContract(address contractAddress) internal {
+        bytes32 resume_role = IPausableUntil(contractAddress).RESUME_ROLE();
+
+        IAccessControlEnumerable(contractAddress).grantRole(resume_role, address(this));
+        IPausableUntil(contractAddress).resume();
+        IAccessControlEnumerable(contractAddress).renounceRole(resume_role, address(this));
+    }
+
+    function _accountingOracle() internal view returns (IAccountingOracle) {
+        return IAccountingOracle(_locator.accountingOracle());
+    }
+
+    function _burner() internal view returns (IBurner) {
+        return IBurner(_locator.burner());
+    }
+
+    function _depositSecurityModule() internal view returns (IDepositSecurityModule) {
+        return IDepositSecurityModule(_locator.depositSecurityModule());
+    }
+
+    function _lido() internal view returns (ILido) {
+        return ILido(_locator.lido());
+    }
+
+    // Returns the same address as _legacyOracle()
+    function _lidoOracle() internal view returns (ILidoOracle) {
+        return ILidoOracle(_locator.legacyOracle());
+    }
+
+    // Returns the same address as _lidoOracle()
+    function _legacyOracle() internal view returns (ILegacyOracle) {
+        return ILegacyOracle(_locator.legacyOracle());
+    }
```

Comment:
Need a comment along the lines of "we're renaming the contract, but it's on the same address"

Reply:
added